File: backend/routes/questions.py
# Copyright (c) 2026 Paul Christopher Cerda
# This source code is licensed under the Business Source License 1.1

"""
Aristotelian observation questions — Block 11

Questions guide students toward evidence-based answers through:
  - Careful observation (What do you see/hear/measure?)
  - Classification (What kind of thing is this?)
  - Causation (What caused this? What will happen next?)
  - Comparison (How does this differ from...?)
  - Evidence (What would prove/disprove this?)

NOT abstract (not "what is justice?") — these have concrete, observable answers grounded in the Aristotelian tradition.
"""

from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text
from pathlib import Path
from typing import Optional
import io
import json
import logging
import sqlite3
import tempfile
import os

from core.database import get_db
from core.dependencies import get_current_user

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/v1/aristotelian-questions", tags=["questions"])


# ── Seed data ─────────────────────────────────────────────────────────────────
# Lives in data/aristotelian_questions.json, not inline here — at 190+ entries
# (up from the original 26) a literal Python list stopped being something
# anyone could reasonably review or edit by hand. Loaded once at import time;
# format is {subject, grade_band, bloom_level, observation_type,
# question_text, follow_up} per entry.

_SEED_PATH = Path(__file__).parent.parent / "data" / "aristotelian_questions.json"


def _load_seed_questions() -> list[tuple]:
    with open(_SEED_PATH, encoding="utf-8") as f:
        rows = json.load(f)
    return [
        (r["subject"], r["grade_band"], r["bloom_level"], r["observation_type"], r["question_text"], r["follow_up"])
        for r in rows
    ]


SEED_QUESTIONS = _load_seed_questions()


# ── Startup table creation ────────────────────────────────────────────────────

# asyncpg rejects multiple statements in a single execute() — split into separate calls
CREATE_TABLE_SQL = """
CREATE TABLE IF NOT EXISTS aristotelian_questions (
    id              SERIAL PRIMARY KEY,
    subject         VARCHAR(50)  NOT NULL,
    grade_band      VARCHAR(10)  NOT NULL,
    bloom_level     VARCHAR(20)  NOT NULL,
    observation_type VARCHAR(30) NOT NULL,
    question_text   TEXT        NOT NULL,
    follow_up       TEXT,
    created_at      TIMESTAMP   DEFAULT NOW()
)
"""
CREATE_INDEXES_SQL = [
    "CREATE INDEX IF NOT EXISTS idx_aq_subject ON aristotelian_questions(subject)",
    "CREATE INDEX IF NOT EXISTS idx_aq_grade   ON aristotelian_questions(grade_band)",
    "CREATE INDEX IF NOT EXISTS idx_aq_bloom   ON aristotelian_questions(bloom_level)",
    # Backs ON CONFLICT (question_text) below — question_text is the natural
    # key here (there's no other stable identifier in the seed JSON), so a
    # unique index on it is what makes reseeding idempotent.
    "CREATE UNIQUE INDEX IF NOT EXISTS idx_aq_question_text_unique ON aristotelian_questions(question_text)",
]

SEED_SQL = """
INSERT INTO aristotelian_questions
    (subject, grade_band, bloom_level, observation_type, question_text, follow_up)
VALUES (:subject, :grade_band, :bloom_level, :observation_type, :question_text, :follow_up)
ON CONFLICT (question_text) DO NOTHING
"""


async def ensure_questions_table(engine):
    """
    Create table/indexes and seed any new questions — called from main.py
    startup on every boot, not just when the table is empty. The seed JSON
    started at 26 questions and grew to 190+; gating the insert loop behind
    `count == 0` meant an already-seeded dev database from the 26-question
    era would never pick up the rest without a manual wipe. ON CONFLICT
    (question_text) DO NOTHING (backed by the unique index above) makes
    re-running the full seed list on every startup cheap and safe — already-
    present rows are skipped, only genuinely new ones get inserted.
    """
    try:
        async with engine.begin() as conn:
            await conn.execute(text(CREATE_TABLE_SQL))
            for idx_sql in CREATE_INDEXES_SQL:
                await conn.execute(text(idx_sql))
            before = (await conn.execute(text("SELECT COUNT(*) FROM aristotelian_questions"))).scalar()
            for q in SEED_QUESTIONS:
                await conn.execute(text(SEED_SQL), dict(zip(
                    ["subject", "grade_band", "bloom_level", "observation_type", "question_text", "follow_up"], q
                )))
            after = (await conn.execute(text("SELECT COUNT(*) FROM aristotelian_questions"))).scalar()
            if after > before:
                logger.info(f"✅ Aristotelian questions: {after - before} new question(s) seeded ({after} total)")
            else:
                logger.info(f"✅ Aristotelian questions table ready ({after} questions)")
    except Exception as e:
        logger.warning(f"⊘ Questions table setup skipped: {e}")


# ── API endpoint ──────────────────────────────────────────────────────────────

@router.get("")
async def get_questions(
    subject: Optional[str] = Query(None),
    grade: Optional[str] = Query(None),
    bloom: Optional[str] = Query(None),
    observation_type: Optional[str] = Query(None),
    limit: int = Query(5, le=20),
    db: AsyncSession = Depends(get_db),
):
    """
    Return Aristotelian observation questions filtered by subject, grade band, and bloom level.
    Used by student capture flow to surface relevant evidence-gathering questions.
    """
    conditions = []
    params: dict = {"limit": limit}

    if subject:
        conditions.append("subject = :subject")
        params["subject"] = subject.lower()
    if grade:
        conditions.append("grade_band = :grade")
        params["grade"] = grade.lower()
    if bloom:
        conditions.append("bloom_level = :bloom")
        params["bloom"] = bloom.lower()
    if observation_type:
        conditions.append("observation_type = :obs_type")
        params["obs_type"] = observation_type.lower()

    where = ("WHERE " + " AND ".join(conditions)) if conditions else ""
    sql = f"SELECT * FROM aristotelian_questions {where} ORDER BY RANDOM() LIMIT :limit"

    try:
        result = await db.execute(text(sql), params)
        rows = result.mappings().all()
        return {"questions": [dict(r) for r in rows], "count": len(rows)}
    except Exception as e:
        logger.error(f"Questions query error: {e}")
        return {"questions": [], "count": 0}


@router.get("/subjects")
async def list_subjects(db: AsyncSession = Depends(get_db)):
    """List all available subjects."""
    try:
        result = await db.execute(text(
            "SELECT DISTINCT subject FROM aristotelian_questions ORDER BY subject"
        ))
        return {"subjects": [r[0] for r in result.fetchall()]}
    except Exception:
        return {"subjects": ["science", "math", "history", "art", "language"]}


@router.get("/sqlite", response_class=StreamingResponse)
async def download_sqlite(db: AsyncSession = Depends(get_db)):
    """
    Return all questions as a self-contained SQLite file (Block 11.4).
    Mobile downloads this once on first launch and queries locally thereafter.
    No auth required — questions are not sensitive.
    """
    try:
        result = await db.execute(text(
            "SELECT id, subject, grade_band, bloom_level, observation_type, "
            "question_text, follow_up FROM aristotelian_questions ORDER BY id"
        ))
        rows = result.fetchall()
    except Exception as e:
        logger.error(f"SQLite export query error: {e}")
        rows = []

    # Build SQLite in a temp file then stream it
    tmp_path = f"/tmp/questions_{id(rows)}.sqlite"
    try:
        conn = sqlite3.connect(tmp_path)
        cur = conn.cursor()
        cur.execute("""
            CREATE TABLE IF NOT EXISTS aristotelian_questions (
                id               INTEGER PRIMARY KEY,
                subject          TEXT NOT NULL,
                grade_band       TEXT NOT NULL,
                bloom_level      TEXT NOT NULL,
                observation_type TEXT NOT NULL,
                question_text    TEXT NOT NULL,
                follow_up        TEXT
            )
        """)
        cur.execute("CREATE INDEX IF NOT EXISTS idx_subject ON aristotelian_questions(subject)")
        cur.execute("CREATE INDEX IF NOT EXISTS idx_grade   ON aristotelian_questions(grade_band)")
        cur.execute("CREATE INDEX IF NOT EXISTS idx_bloom   ON aristotelian_questions(bloom_level)")
        if rows:
            cur.executemany(
                "INSERT OR IGNORE INTO aristotelian_questions "
                "(id, subject, grade_band, bloom_level, observation_type, question_text, follow_up) "
                "VALUES (?,?,?,?,?,?,?)",
                rows
            )
        else:
            # DB empty — fall back to in-memory seed constants
            for i, q in enumerate(SEED_QUESTIONS, start=1):
                cur.execute(
                    "INSERT OR IGNORE INTO aristotelian_questions "
                    "(id, subject, grade_band, bloom_level, observation_type, question_text, follow_up) "
                    "VALUES (?,?,?,?,?,?,?)",
                    (i, *q)
                )
        conn.commit()
        conn.close()

        buf = io.BytesIO()
        with open(tmp_path, "rb") as f:
            buf.write(f.read())
    finally:
        os.unlink(tmp_path)

    buf.seek(0)
    return StreamingResponse(
        buf,
        media_type="application/octet-stream",
        headers={"Content-Disposition": "attachment; filename=questions.sqlite"},
    )

File: mobile/src/db/questions.ts
// src/db/questions.ts
// Downloads questions.sqlite from backend on first launch (or when stale),
// then queries locally. Zero network required in the field.

import * as FileSystem from 'expo-file-system';
import { getDb } from './database';
import { API_BASE, getToken } from '@/src/api/client';

const QUESTIONS_CACHE_DAYS = 7;

export interface LocalQuestion {
  id: number;
  subject: string;
  grade_band: string;
  bloom_level: string;
  observation_type: string;
  question_text: string;
  follow_up: string | null;
}

/** Returns true if local question DB is empty or older than QUESTIONS_CACHE_DAYS */
async function needsRefresh(): Promise<boolean> {
  const db = await getDb();
  const count = await db.getFirstAsync<{ n: number }>('SELECT COUNT(*) as n FROM questions');
  if (!count || count.n === 0) return true;

  const meta = await db.getFirstAsync<{ value: string }>(
    "SELECT value FROM questions_meta WHERE key = 'downloaded_at'"
  );
  if (!meta) return true;

  const ageMs = Date.now() - parseInt(meta.value, 10);
  return ageMs > QUESTIONS_CACHE_DAYS * 86_400_000;
}

/** Download questions.sqlite from backend and import into local DB */
export async function syncQuestionsFromServer(): Promise<void> {
  try {
    const token = await getToken();
    const headers: Record<string, string> = token
      ? { Authorization: `Bearer ${token}` }
      : {};

    // Download the sqlite file
    const tmpPath = FileSystem.cacheDirectory + 'questions_download.sqlite';
    const result = await FileSystem.downloadAsync(
      `${API_BASE}/api/v1/aristotelian-questions/sqlite`,
      tmpPath,
      { headers }
    );
    if (result.status !== 200) throw new Error(`Download failed: ${result.status}`);

    // Open downloaded DB and read all rows
    const { SQLiteDatabase } = await import('expo-sqlite');
    const downloadedDb = await (await import('expo-sqlite')).openDatabaseAsync(tmpPath);
    const rows = await downloadedDb.getAllAsync<LocalQuestion>(
      'SELECT * FROM aristotelian_questions ORDER BY id'
    );
    await downloadedDb.closeAsync();
    await FileSystem.deleteAsync(tmpPath, { idempotent: true });

    if (rows.length === 0) return; // Nothing to import

    // Insert into local DB
    const db = await getDb();
    await db.runAsync('DELETE FROM questions');
    await db.withTransactionAsync(async () => {
      for (const q of rows) {
        await db.runAsync(
          `INSERT OR REPLACE INTO questions
           (id, subject, grade_band, bloom_level, observation_type, question_text, follow_up)
           VALUES (?, ?, ?, ?, ?, ?, ?)`,
          [q.id, q.subject, q.grade_band, q.bloom_level, q.observation_type, q.question_text, q.follow_up ?? null]
        );
      }
      await db.runAsync(
        "INSERT OR REPLACE INTO questions_meta (key, value) VALUES ('downloaded_at', ?)",
        [Date.now().toString()]
      );
    });

    console.log(`✅ Questions synced: ${rows.length} questions stored locally`);
  } catch (e) {
    console.warn('⊘ Questions sync failed (will use cached or API fallback):', e);
  }
}

/** Get a random question from local DB — zero network. Falls back to null. */
export async function getLocalQuestion(params: {
  subject?: string;
  gradeBand?: string;
  bloomLevel?: string;
}): Promise<LocalQuestion | null> {
  try {
    const db = await getDb();
    const conditions: string[] = [];
    const args: string[] = [];

    if (params.subject) {
      conditions.push('subject = ?');
      args.push(params.subject.toLowerCase());
    }
    if (params.gradeBand) {
      conditions.push('grade_band = ?');
      args.push(params.gradeBand.toLowerCase());
    }
    if (params.bloomLevel) {
      conditions.push('bloom_level = ?');
      args.push(params.bloomLevel.toLowerCase());
    }

    const where = conditions.length ? 'WHERE ' + conditions.join(' AND ') : '';
    const row = await db.getFirstAsync<LocalQuestion>(
      `SELECT * FROM questions ${where} ORDER BY RANDOM() LIMIT 1`,
      args
    );
    return row ?? null;
  } catch {
    return null;
  }
}

/** Initialise questions on app start — refresh if stale */
export async function initQuestions(): Promise<void> {
  if (await needsRefresh()) {
    await syncQuestionsFromServer();
  }
}
